[Merged by Bors] - feat(analysis/complex/upper_half_plane/metric): prove that SL(2, ℝ) acts isometrically on upper half space with the hyperbolic metric

RuizheWan opened the pull request:
A key part of the argument is to show that the element `modular_group.S` acts isometrically. We thus move the definition `modular_group.S` (and its partner `modular_group.T`) earlier in the import hierarchy to make it available without introducing a dependency on the theory of the modular group.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

RuizheWan committed:
isometry_on_upper_half_plane

RuizheWan committed:
some message

ocfnash committed:
Whitespace tweaks

ocfnash committed:
Slightly reorganise simp lemmas to fix lint errors

ocfnash committed:
Drop duplicate `open_locale` statement.

ocfnash committed:
Reorder lemmas slightly

ocfnash committed:
Tidy up and golf a little

There is still a lot to be done inside proofs but these changes leave the
external API is reasonable shape.

ocfnash committed:
Fix lint error

ocfnash committed:
Cosmetic tweaks

ocfnash committed:
Golf proof of `isometry_involute`.

ocfnash committed:
Some more golfing

ocfnash reviewed (changes requested):
I think these changes don't need to go in a new file. At least half of them can go in `analysis/complex/upper_half_plane/metric.lean` and the rest have natural homes too.

ocfnash committed:
Fold proof of `isometry_involute` into `isometry_SL2_smul`.

Since `isometry_involute` is a special case, it will never be needed after
proving `isometry_involute`.

Other minor golfs.

RuizheWan committed:
placing the lemmas in the right place.

RuizheWan committed:
Placing the lemmas in the right place.

RuizheWan committed:
placing the lemmas in the right place.

RuizheWan committed:
Correcting some errors

RuizheWan committed:
Modifying line length

ocfnash committed:
Reorganise slightly

eric-wieser commented on src/analysis/complex/upper_half_plane/basic.lean:
```diff
@@ -65,7 +65,9 @@ def mk (z : ℂ) (h : 0 < z.im) : ℍ := ⟨z, h⟩
 @[simp] lemma coe_re (z : ℍ) : (z : ℂ).re = z.re := rfl
 
 @[simp] lemma mk_re (z : ℂ) (h : 0 < z.im) : (mk z h).re = z.re := rfl
+@[simp] lemma mk_re' (z : ℂ) (h) : re (⟨z, h⟩ : ℍ) = z.re := rfl
```
Which `re` is this on the LHS? (I'll check when I'm next at lean if you're not)

RuizheWan replied:
This is `def re (z : ℍ) := (z : ℂ).re` at line 57.

eric-wieser replied:
Then isn't this exactly the same as the lemma above?

eric-wieser replied:
Ah, I see: these are about `subtype.mk` not `mk`. I still think we probably don't want these lemmas; do you actually use them in this PR?

RuizheWan replied:
Yes I think `mk_re'` is used in `simp` somewhere.

eric-wieser replied:
Can you remove it anyway? If it's being used by simp, then you have a proof somewhere that is unfolding things it shouldn't

ocfnash replied:
These were used in a previous version of some proof(s) which unfolded all the way to `ℂ` instead of staying in `ℍ`. 

Personally I think it's fine to keep them but I I've pushed a commit which drops them since they are not in fact needed any more.

eric-wieser commented on src/analysis/complex/upper_half_plane/metric.lean:
```diff
@@ -349,4 +349,25 @@ begin
 end
 
+/-- `SL(2, ℝ)` acts on the upper half plane as an isometry.-/
+lemma isometry_SL2_smul (g : SL(2, ℝ)) : isometry ((•) g : ℍ → ℍ) :=
```
I think we now have an `isometric_smul` typeclass?

eric-wieser commented on src/linear_algebra/matrix/special_linear_group.lean:
```diff
@@ -232,6 +232,31 @@ begin
   refl,
 end
 
+lemma fin_two_exists_eq_mk (g : SL(2, R)) :
```
I think this would be more powerful if stated as an induction principle, something like
```
(h : ∀ a b c d (hdet : a * d - b * c = 1), P ⟨!![a, b; c, d], by rwa [matrix.det_fin_two_of]⟩) : ∀ x, P x
```

ocfnash replied:
Good point, I guess the lemma would be:
```lean
lemma fin_two_induction (P : SL(2, R) → Prop)
  (h : ∀ a b c d (hdet : a * d - b * c = (1 : R)), P ⟨!![a, b; c, d], by rwa [det_fin_two_of]⟩)
  (g : SL(2, R)) : P g :=
begin
  obtain ⟨m, hm⟩ := g,
  convert h (m 0 0) (m 0 1) (m 1 0) (m 1 1) (by rwa det_fin_two at hm),
  ext i j, fin_cases i; fin_cases j; refl,
end
```

eric-wieser replied:
It would be even easier if we had #15738

eric-wieser replied:
`∀ (a b c d : R) (hdet : a * d - b * c = 1)` without the type ascription on `1` feels slightly clearer to me.

ocfnash replied:
I have added the lemma `fin_two_induction` but I have kept `fin_two_exists_eq_mk` based on the fact that I would otherwise need the following diff in the proof of `exists_SL2_smul_eq_of_apply_zero_one_ne_zero`:
```diff
-   obtain ⟨a, b, c, d, h, rfl⟩ := g.fin_two_exists_eq_mk,
+   revert h_denom hc,
+   refine fin_two_induction _ (λ a b c d h, _) g,
+   intros h_denom hc,
```
which persuades me that it's worth having both.

eric-wieser replied:
Can't you do `induction g using the_namespace.fin_two_induction`?

ocfnash replied:
Ah yes.

I always forget about this feature of the `induction` tactic.

Thanks!

eric-wieser commented on src/number_theory/modular.lean:
```diff
@@ -314,16 +314,6 @@ begin
-
-lemma coe_T : ↑ₘT = !![1, 1; 0, 1] := rfl
-
 lemma coe_T_inv : ↑ₘ(T⁻¹) = !![1, -1; 0, 1] := by simp [coe_inv, coe_T, adjugate_fin_two]
```
Does it make sense to move these too?

ocfnash replied:
Yeah it probably does. E.g., everything up to `im_T_inv_smul`

ocfnash replied:
I've moved more additional lemmas but had to exclude `coe_T_zpow_smul_eq` and those below because they need the action on `ℍ`.

At some point (after the port) I'd like to tidy up this corner of the codebase.

eric-wieser reviewed (commented):
I'll give this another round of review once I'm not on mobile.

RuizheWan committed:
making isometry_SL2_smul an instance

eric-wieser committed:
replace non-simp-normal coercions with simp-normal ones

eric-wieser committed:
reduce lambdas

eric-wieser commented on src/analysis/complex/basic.lean:
```diff
@@ -116,6 +116,15 @@ by rw [dist_comm, dist_conj_self]
+  rw [eq_div_iff h, dist_eq, dist_comm, dist_eq, ← absolute_value.map_mul,
+    ← absolute_value.map_mul, sub_mul, inv_mul_cancel_left₀ hz, mul_comm z,
+    inv_mul_cancel_left₀ hw],
+end
```
A shorter proof:
```suggestion
protected lemma dist_inv_inv (z w : ℂ) (hz : z ≠ 0) (hw : w ≠ 0) :
  dist z⁻¹ w⁻¹ = (dist z w) / (abs z * abs w) :=
begin
  rw [dist_eq_norm, inv_sub_inv hz hw, norm_div, dist_eq_norm', norm_mul],
  refl,
end
```

RuizheWan committed:
Update src/analysis/complex/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/analysis/complex/basic.lean:
```diff
@@ -116,6 +116,13 @@ by rw [dist_comm, dist_conj_self]
 lemma nndist_self_conj (z : ℂ) : nndist z (conj z) = 2 * real.nnabs z.im :=
 by rw [nndist_comm, nndist_conj_self]
 
+protected lemma dist_inv_inv (z w : ℂ) (hz : z ≠ 0) (hw : w ≠ 0) :
```
This statement is true for any normed_field, right?

ocfnash replied:
Good point, yes. The following should probably go about [here](https://github.com/leanprover-community/mathlib/blob/ef7acf407d265ad4081c8998687e994fa80ba70c/src/analysis/normed/field/basic.lean#L578): 
```lean
import analysis.normed.field.basic

namespace normed_field

variables {α : Type*} [normed_field α]

protected lemma dist_inv_inv (z w : α) (hz : z ≠ 0) (hw : w ≠ 0) :
  dist z⁻¹ w⁻¹ = (dist z w) / (‖z‖ * ‖w‖) :=
by rw [dist_eq_norm, inv_sub_inv hz hw, norm_div, dist_eq_norm', norm_mul]

end normed_field
```

eric-wieser replied:
A `normed_division_ring` version would be nice too (using the primed `inv_sub_inv'`)

eric-wieser replied:
In fact I assume the statement would end  up the same, since the division is in `real`; so we only need one lemma, but proved in the more general case.

eric-wieser replied:
we should have the corresponding `nndist` lemma too.

ocfnash committed:
Drop `simp` lemmas based on `subtype.mk`.

sgouezel commented on src/analysis/complex/upper_half_plane/basic.lean:
```diff
@@ -30,11 +30,13 @@ open_locale classical big_operators matrix_groups
 local attribute [instance] fintype.card_fin_even
 
-/- Disable this instances as it is not the simp-normal form, and having them disabled ensures
+/- Disable these instances as they is not the simp-normal form, and having them disabled ensures
```
```suggestion
/- Disable these instances as they are not the simp-normal form, and having them disabled ensures
```

RuizheWan committed:
Update src/analysis/complex/upper_half_plane/basic.lean

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

ocfnash committed:
Generalise `complex.dist_inv_inv` to any `normed_division_ring`.

eric-wieser commented on src/analysis/normed/field/basic.lean:
```diff
@@ -453,6 +453,15 @@ nnreal.eq $ by simp
 @[simp] lemma nnnorm_zpow : ∀ (a : α) (n : ℤ), ‖a ^ n‖₊ = ‖a‖₊ ^ n :=
 map_zpow₀ (nnnorm_hom : α →*₀ ℝ≥0)
 
+protected lemma normed_division_ring.dist_inv_inv {z w : α} (hz : z ≠ 0) (hw : w ≠ 0) :
```
I think either `dist_inv_inv'` or `dist_inv_inv\0` would be a better name

ocfnash replied:
Agreed!

ocfnash committed:
Rename `normed_division_ring.dist_inv_inv` to `dist_inv_inv₀`.

ocfnash committed:
Fix typo

ocfnash committed:
Add lemma `special_linear_group.fin_two_induction`

ocfnash committed:
Move more lemmas from `number_theory.modular` to `linear_algebra.matrix.special_linear_group`.

ocfnash committed:
Drop `fin_two_exists_eq_mk` in favour of `fin_two_induction`.

Also golf proof of `fin_two_exists_eq_mk_of_apply_zero_one_eq_zero`

eric-wieser commented on src/analysis/complex/upper_half_plane/basic.lean:
```diff
@@ -330,7 +330,7 @@ lemma exists_SL2_smul_eq_of_apply_zero_one_ne_zero (g : SL(2, ℝ)) (hc : ↑ₘ
 begin
   have h_denom := denom_ne_zero g,
-  obtain ⟨a, b, c, d, h, rfl⟩ := g.fin_two_exists_eq_mk,
+  induction g using matrix.special_linear_group.fin_two_induction with a b c d h_det,
```
Any reason for renaming from `h` to `h_det`? Won't the proof below need to change too?

ocfnash replied:
This was accidental. I got away with it because of the "feature" of `replace` (used below) which makes it behave like `have` if the hypothesis did not actually exist.

In any case, I've reverted the rename now.

ocfnash committed:
Revert rename

jcommelin reviewed (approved):
Thanks :tada:

bors merge
